closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -362,6 +376,42 @@ int encryptFinal(byte[] in, int inOfs, int len, byte[] out, int outOfs)
         return (len + tagLenBytes);
     }
 
+    /**
+     * Performs encryption operation for the last time.
+     *
+     * @param src the input buffer with the data to be encrypted
+     * @param dst the offset in <code>in</code>
+     * @return the number of bytes placed into the <code>out</code> buffer
+     */
+    int encryptFinal(ByteBuffer src, ByteBuffer dst)
+        throws IllegalBlockSizeException, ShortBufferException {
+        byte[] dst_arr;
+
+        // Get array from source
+        byte[] src_arr = getBbArray(src);
+        int src_ofs = src.position() + (src.hasArray() ? src.arrayOffset() : 0);
+
+        // Get array from destination
+        if (dst.hasArray()) {
+            dst_arr = dst.array();
+        } else {
+            dst_arr = new byte[dst.capacity()];
+        }
+        int dst_offset = dst.position() + (dst.hasArray() ? dst.arrayOffset() : 0);
+
+        int len = encryptFinal(src_arr, src_ofs, src.remaining(), dst_arr, dst_offset);
+
+        // Update destination buffer
+        if (dst.hasArray()) {
+            dst.position(dst.position() + len);
+            dst.limit(dst.position());
+        } else {
+            dst.put(dst_arr, dst.position(), len);
+        }
```

Comment:
The destination array does not need to be so large; it need be no larger than `dst.remaining()`.
When a new array is created, `dst_offset` should be zero.
The position of the source buffer must be updated to reflect consumption of input bytes.
The destination buffer limit should not be modified.
```
        // Get array from source
        byte[] src_arr = getBbArray(src);
        int src_ofs = src.hasArray() ? (src.position() + src.arrayOffset()) : 0;

        // Get array from destination
        byte[] dst_arr;
        int dst_offset;

        if (dst.hasArray()) {
            dst_arr = dst.array();
            dst_offset = dst.position() + dst.arrayOffset();
        } else {
            dst_arr = new byte[src.remaining()];
            dst_offset = 0;
        }

        int len = encryptFinal(src_arr, src_ofs, src.remaining(), dst_arr, dst_offset);

        // Advance source buffer position
        src.position(src.position() + len); // FIXME len may be too large

        // Update destination buffer content and position
        if (dst.hasArray()) {
            dst.position(dst.position() + len);
        } else {
            dst.put(dst_arr, 0, len);
        }
```

Reply:
Shouldn't the src position become the src limit?
Because we are reading all of the `src.remaining()` and encrypting them. So I think we could simple add afterwards:
```java
src.position(src.limi());
```

Reply:
Yes, that's reasonable assuming `encryptFinal` has succeeded and consumed all the bytes it was given.

Reply:
Done.